public/app/plugins/datasource/loki/components/monaco-query-field/MonacoQueryField.tsx

```diff
@@ -206,6 +213,32 @@ const MonacoQueryField = ({ history, onBlur, onRunQuery, initialValue, datasourc
               editor.trigger('', 'editor.action.triggerSuggest', {});
             }
           });
+
+          const placeholderDecorators = [
+            {
+              range: new monaco.Range(1, 1, 1, 1),
+              options: {
+                className: styles.placeholder,
+                isWholeLine: true,
+              },
+            },
+          ];
+
+          let decorators: string[] = [];
+
+          const checkDecorators: () => void = () => {
+            const model = editor.getModel();
+
+            if (!model) {
+              return;
+            }
+
+            const newDecorators = model.getValueLength() === 0 ? placeholderDecorators : [];
+            decorators = model.deltaDecorations(decorators, newDecorators);
+          };
+
+          checkDecorators();
+          editor.onDidChangeModelContent(checkDecorators);
```

Comment:
Could we wrap this logic into a function and then just call `setPlaceholder(editor, monaco)` in here. In the future, it is going to be easier to understand which parts are actually setting placeholder and which not. WDYT?

Reply:
Because now the `onMount` has already ~100 lines of things to run. So knowing what is what and why might be beneficial in the future.

Reply:
yeah that makes sense! i can update that 😄